File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
import sys
input = sys.stdin.readline

N, M, V = map(int, input().split())
graph = {i : [] for i in range(N+1)}

rs_dfs = [V]
rs_bfs = [V]
for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
for i in range(N+1):
    graph[i].sort()

visited = [False for _ in range(N+1)]
visited[V] = True
def dfs(v):
    for next in graph[v]:
        if visited[next]:
            continue
        visited[next] = True
        rs_dfs.append(next)
        dfs(next)

def bfs(v):
    visited = [False for _ in range(N+1)]
    visited[V] = True
    q = deque([v])
    while q:
        cur = q.popleft()
        for next in graph[cur]:
            if visited[next]:
                continue
            q.append(next)
            visited[next] = True
            rs_bfs.append(next)

dfs(V)
bfs(V)

for e in rs_dfs:
    print(e, end=' ')
print()
for e in rs_bfs:
    print(e, end=' ')

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
    private static StringBuilder sb = new StringBuilder();
    private static int[][] graph;
    private static boolean[] visited;
    private static int N;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int V = Integer.parseInt(st.nextToken());
        graph = new int[N+1][N+1];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a][b] = graph[b][a] = 1;
        }

        visited = new boolean[N+1];
        dfs(V);
        sb.append("\n");
        visited = new boolean[N+1];
        bfs(V);
        sb.append("\n");
        bw.write(sb.toString());

        bw.flush();
        br.close();
        bw.close();
    }

    private static void dfs(int node) {
        sb.append(node + " ");
        visited[node] = true;
        for (int next = 1; next <= N; next++) {
            if (graph[node][next] == 1 && !visited[next]) {
                dfs(next);
            }
        }
    }

    private static void bfs(int node) {
        Queue<Integer> q = new LinkedList<>();
        q.add(node);
        visited[node] = true;

        while(!q.isEmpty()) {
            int cur = q.poll();
            sb.append(cur + " ");
            for (int next = 1; next <= N; next++) {
                if (graph[cur][next] == 1 && !visited[next]) {
                    q.add(next);
                    visited[next] = true;
                }
            }
        }
    }
}